runtime/interpreter/interpreter.go

```diff
@@ -2214,8 +2217,182 @@ func (interpreter *Interpreter) writeStored(
+		},
+		functionType,
+	)
+	return FromStringFunctionValue{
+		parentType:   ty,
+		hostFunction: hostFunctionImpl,
+	}
+}
+
+// default implementation for parsing a given unsigned numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings must not begin with a '+' or '-'.
+func fromStringImplUnsigned(bitSize int, convert func(uint64) Value) func(*Interpreter, string) Value {
+	expectedUsage := common.NewNumberMemoryUsage(bitSize / 8)
+
+	return func(interpreter *Interpreter, input string) Value {
+		val, err := strconv.ParseUint(input, 10, bitSize)
+		if err != nil {
+			return NewNilValue(interpreter)
+		}
+
+		converted := convert(val)
+		common.UseMemory(interpreter, expectedUsage)
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+// default implementation for parsing a given signed numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings may begin with a '+' or '-'.
+func fromStringImplSigned(bitSize int, convert func(int64) Value) func(*Interpreter, string) Value {
+	expectedUsage := common.NewCadenceBigIntMemoryUsage(bitSize / 8)
+
+	return func(interpreter *Interpreter, input string) Value {
+		val, err := strconv.ParseInt(input, 10, bitSize)
+		if err != nil {
+			return NewNilValue(interpreter)
+		}
+
+		converted := convert(val)
+		common.UseMemory(interpreter, expectedUsage)
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+func fromStringImplBig(convert func(*big.Int) (Value, bool)) func(*Interpreter, string) Value {
+	return func(interpreter *Interpreter, input string) Value {
+		val, ok := common.NewBigInt(interpreter).SetString(input, 10)
+		if !ok {
+			return NewNilValue(interpreter)
+		}
+		curSize := len(val.Bits())
+		common.UseMemory(interpreter, common.NewCadenceBigIntMemoryUsage(curSize))
+
+		converted, ok := convert(val)
+
+		if !ok {
+			return NewNilValue(interpreter)
+		}
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+// check if val is in the inclusive interval [low, high]
+func inRange(val *big.Int, low *big.Int, high *big.Int) bool {
+	return -1 < val.Cmp(low) && val.Cmp(high) < 1
+}
+
+var FromStringFunctionValues = func() map[string]FromStringFunctionValue {
+	declarations := []FromStringFunctionValue{
+		fromStringDeclaration(sema.Int8Type, fromStringImplSigned(8, func(n int64) Value {
+			return NewUnmeteredInt8Value(int8(n))
+		})),
+		fromStringDeclaration(sema.Int16Type, fromStringImplSigned(16, func(n int64) Value {
+			return NewUnmeteredInt16Value(int16(n))
+		})),
+		fromStringDeclaration(sema.Int32Type, fromStringImplSigned(32, func(n int64) Value {
+			return NewUnmeteredInt32Value(int32(n))
+		})),
+		fromStringDeclaration(sema.Int64Type, fromStringImplSigned(64, func(n int64) Value {
+			return NewUnmeteredInt64Value(n)
```

Comment:
All these methods should use the 'metered' version of the constructors. Because one can create a lot of numbers from its string representation inside a loop, and those needs to be metered.

Reply:
the metering is also done in the fromString- wrapper functions to avoid redundancy

Reply:
Oh, I see. 
The 'metered' constructor guarantees consistency. i.e: all IntX types are metered in one place. So if we have to update at some point, then only have to update one place / wont miss out to update other places.

Also, the fromString- wrapper first creates the value - and then is metered. The constructor actually does it the other way around (so the limit is never exceeded).

Reply:
after a LOT of generics abuse, i've managed to fix things to use metered constructors